src/main/java/fish/payara/extras/upgrade/BaseUpgradeCommand.java

```diff
@@ -275,9 +276,34 @@ protected void reinstallNodes() throws IOException, CommandException, Configurat
                         throwException = true;
                         failingNodes.add(node.getName());
                     }
+                } else if (node.getType().equals("CONFIG")) {
+                    // For local instances, remove osgi-cache content
+                    foundNode = true;
+                    String nodeDir = node.getInstallDir();
+                    // nodeDir can contain variable, evaluate it
+                    nodeDir = nodeDir.replaceAll("\\$\\{com.sun.aas.productRoot\\}", System.getProperty("com.sun.aas.productRoot"));
+                    Path nodePath = Paths.get(nodeDir, "glassfish", "nodes", node.getName());
+                    if (Files.exists(nodePath)) {
+                        List<Exception> exceptions = Files.list(nodePath)
+                                .map(p -> p.resolve("osgi-cache")) // interested in osgi-cache subdirectories
+                                .filter(p -> p.toFile().exists()) // only if it exists
+                                .peek(p -> logger.fine(String.format("Deleting osgi-cache for local instance, path %s", p)))
+                                .flatMap(p -> Stream.of(p.toFile().listFiles())) // process subdirectories, not osgi-cache itself
+                                .map(f -> f.toPath())
+                                .map(p -> deleteDirectoryRecursively(p)) // delete all the directories
+                                .filter(e -> e != null)
+                                .collect(Collectors.toList()); // collect all exceptions
+                        if (!exceptions.isEmpty()) {
+                            exceptions.stream().forEach(e -> logger.log(Level.SEVERE, e.getMessage(), e));
+                            throwException = true;
+                            failingNodes.add(node.getName());
+                        }
+                    } else {
+                        logger.fine(String.format("Node directory %s not found, skipping.", nodePath));
+                    }
```

Comment:
There's a minor bug with this due to it attempting to do this on remote CONFIG nodes.

If you have a domain (e.g. domain1) on a machine (machine1) with a remote CONFIG node on another machine (machine2) named noddy, and then a second domain (production) on the same machine (machine1) with a _local_ CONFIG node (so also on machine1) with the same name (noddy), the script will attempt to delete the OSGi cache of noddy twice.

```
Reinstalling nodes for domain domain1
Node directory /home/andrew/payara5/glassfish/nodes/localhost-domain1 not found, skipping.
Deleting osgi-cache for local instance, path /home/andrew/payara5/glassfish/nodes/noddy/Old-fashioned-Devario/osgi-cache
Reinstalling nodes for domain production
Node directory /home/andrew/payara5/glassfish/nodes/localhost-production not found, skipping.
Deleting osgi-cache for local instance, path /home/andrew/payara5/glassfish/nodes/noddy/Old-fashioned-Devario/osgi-cache
Command reinstall-nodes executed successfully.
```

Reply:
Resolved - deleting only on local node.